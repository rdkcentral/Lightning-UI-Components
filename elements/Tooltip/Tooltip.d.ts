import Base from '../../Base';
import { WithThemeStylesConstructor } from '../../mixins/withThemeStyles';

type transitionObject = {
  delay: number;
  duration: number;
  timingFunction: string;
};

export interface TooltipStyles {
  radius: number;
  textStyle: string;
  marginBottom: number;
  paddingY: number;
  paddingX: number;
  pointerW: number;
  pointerH: number;
  transition: transitionObject;
  backgroundColor: string;
  textColor: string;
}

declare const Tooltip_base: WithThemeStylesConstructor<typeof Base, Tooltip>;

export default class Tooltip extends Base {
  title?: string;
  delayVisible?: number;
  timeVisible?: number;
}
