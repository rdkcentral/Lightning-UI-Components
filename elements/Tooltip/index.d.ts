import Base from '../../Base';
import { WithThemeStylesConstructor } from '../../mixins/withThemeStyles';
import lng from '@lightningjs/core';

type transitionObject = {
  delay: number;
  duration: number;
  timingFunction: string;
};

export interface TooltipStyles {
  backgroundColor: string;
  marginBottom: number;
  paddingX: number;
  paddingY: number;
  pointerW: number;
  pointerH: number;
  radius?: lng.Tools.CornerRadius;
  textStyle: string;
  transition: transitionObject;
}

declare const Tooltip_base: WithThemeStylesConstructor<typeof Base, Tooltip>;

export default class Tooltip extends Base {
  title?: string;
  delayVisible?: number;
  timeVisible?: number;
}
