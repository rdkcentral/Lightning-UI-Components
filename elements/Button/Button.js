/**
 * Button Component
 *
 * Generic Button Component that handles setting text icon and checkbox
 */
import { withExtensions } from '../../mixins';
import Surface from '../Surface';
import TextBox from '../TextBox';
import * as styles from './Button.styles';
import { measureTextWidth } from '../../utils';
import Row from '../../layout/Row';

class Button extends Surface {
  static get __componentName() {
    return 'Button';
  }

  static get __themeStyles() {
    return styles;
  }

  static get properties() {
    return ['fixed', 'justify', 'prefix', 'suffix', 'title'];
  }

  static get tags() {
    return [
      ...super.tags,
      'Content',
      { name: 'Title', path: 'Content.Title' },
      { name: 'Prefix', path: 'Content.Prefix' },
      { name: 'Suffix', path: 'Content.Suffix' }
    ];
  }

  static _template() {
    return {
      ...super._template(),
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2
      }
    };
  }

  _update() {
    this._updatePrefix();
    this._updateTitle();
    this._updateSuffix();
    this._updateAllPositioning();
    this._updateTruncation();
  }

  _updateAllPositioning() {
    this._updatePositions();
    this._updateContentDimensions();
    this._updateSurfaceDimensions();
    this._updateContentPosition();
    // delaying the super._update call so we update the surface sizing after the button has finished updating
    super._update();
  }

  $itemChanged() {
    this._updateAllPositioning();
  }

  _onTextBoxChanged() {
    this._updateAllPositioning();
  }

  _updatePositions() {
    if (this._hasPrefix) {
      this._Prefix.x = this._prefixX;
    }
    if (this._hasTitle) {
      this._Title.x = this._titleX;
    }
    if (this._hasSuffix) {
      this._Suffix.x = this._suffixX;
    }
  }

  _updatePrefix() {
    if (this.prefix) {
      let prefixPatch = {
        h: this.style.prefixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Prefix) {
        prefixPatch = {
          ...this._rowProps,
          ...prefixPatch
        };
      }
      this._Content.patch({ Prefix: prefixPatch });
      this._Prefix.items = this._addButtonProps(this.prefix);
    } else {
      this._Content.patch({ Prefix: undefined });
    }
  }

  _updateTitle() {
    if (this._hasTitle) {
      let titlePatch = {
        content: this.title,
        textStyle: {
          textColor: this.style.contentColor,
          ...this.style.textStyle
        }
      };
      if (!this._Title) {
        titlePatch = {
          type: TextBox,
          mountY: 0.5,
          y: h => h / 2,
          signals: {
            textBoxChanged: '_onTextBoxChanged'
          },
          ...titlePatch
        };
      }
      this._Content.patch({ Title: titlePatch });
    } else {
      this.patch({ Content: { Title: undefined } });
    }
  }

  _updateSuffix() {
    if (this.suffix) {
      let suffixPatch = {
        h: this.style.suffixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Suffix) {
        suffixPatch = {
          ...this._rowProps,
          ...suffixPatch
        };
      }
      this._Content.patch({ Suffix: suffixPatch });
      this._Suffix.items = this._addButtonProps(this.suffix);
    } else {
      this._Content.patch({ Suffix: undefined });
    }
  }

  _updateTruncation() {
    /**
     * needs to run before rest of content updates to prevent a bug causing text
     * to sometimes truncate and then un-truncate(with no width change) when
     * toggling fixed after toggling focus
     */
    if (this._Title) {
      this._Title.patch({
        wordWrap: this.fixed,
        wordWrapWidth: this.fixed ? this.w - this._paddingX : 0
      });
    }
  }

  _updateContentDimensions() {
    const contentDimensionsPatch = {};
    const y = this.h / 2;

    // reference transition targets to ensure we aren't checking against intermittent values
    if (this._Content.transition('w').targetValue !== this._contentW) {
      contentDimensionsPatch.w = this._contentW;
    }

    if (this._Content.y !== y) {
      contentDimensionsPatch.y = y;
    }

    if (Object.keys(contentDimensionsPatch).length > 0) {
      if (this._smooth) {
        this._Content.smooth = contentDimensionsPatch;
      } else {
        this._Content.patch(contentDimensionsPatch);
      }
    }
  }

  _updateContentPosition() {
    let mountX;
    let x;

    switch (this.justify) {
      case 'left':
        mountX = 0;
        x = this._paddingLeft;
        break;
      case 'right':
        mountX = 1;
        x = this.w - this._paddingRight;
        break;
      case 'center':
      default:
        mountX = 0.5;
        x = this.w / 2;
        break;
    }

    this._Content.patch({ mountX, x });
  }

  _updateSurfaceDimensions() {
    let newWidth = this.w;
    if (this.fixed) {
      newWidth = this._w;
    } else {
      // if no title, ignore minWidth and use prefix/suffix width
      newWidth =
        !this._hasTitle && (this._hasPrefix || this._hasSuffix)
          ? this._contentW + this._paddingX
          : Math.max(this._contentW + this._paddingX, this.style.minWidth);
    }

    if (newWidth !== this.w) {
      this.w = newWidth;
      this.fireAncestors('$itemChanged');
    }
  }

  _addButtonProps(arr) {
    const items = Array.isArray(arr) ? arr : [arr];
    return items.map(item => ({
      centerInParent: true,
      palette: this.palette,
      mode: this.mode,
      ...item,
      style: {
        color: this.style.contentColor,
        ...item.style
      }
    }));
  }

  _getJustify() {
    return !!this._justify ? this._justify : this.style.justify;
  }

  get _hasPrefix() {
    return this._Prefix && Array.isArray(this.prefix)
      ? this.prefix.length
      : this.prefix;
  }

  get _prefixW() {
    return this._hasPrefix ? this._Prefix.w : 0;
  }

  get _prefixX() {
    return 0;
  }

  get _hasTitle() {
    return !!this.title;
  }

  get _titleW() {
    // get the pre-calculated text width of the title to solve some race condition bugs
    if (
      this._hasTitle &&
      this._Title &&
      this._Title._Text &&
      this._Title.visible
    ) {
      if (this.fixed) {
        return this._Title.w;
      } else {
        return measureTextWidth(this._Title._Text.text);
      }
    }
    return 0;
  }

  get _titleX() {
    return this._hasPrefix ? this._prefixW + this.style.titlePadding : 0;
  }

  get _hasSuffix() {
    return this._Suffix && Array.isArray(this.suffix)
      ? this.suffix.length
      : this.suffix;
  }

  get _suffixW() {
    return this._hasSuffix ? this._Suffix.w : 0;
  }

  get _suffixX() {
    if (this._hasTitle) {
      return this._titleW + this._Title.x + this.style.titlePadding;
    } else if (this._hasPrefix) {
      return this._prefixW + this.style.itemSpacing;
    }
    return 0;
  }

  get _contentW() {
    if (this._hasSuffix) {
      return this._suffixX + this._suffixW;
    } else if (this._hasTitle && this._Title.visible) {
      return this._titleX + this._titleW;
    } else if (this._hasPrefix) {
      return this._prefixX + this._prefixW;
    }
    return 0;
  }

  get _rowProps() {
    return {
      type: Row,
      mountY: 0.5,
      autoResizeWidth: true
    };
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }

    // TODO - Localization?
    // Do we need a locale file with
    // component translations?
    return this.title + ', Button';
  }

  get _paddingX() {
    return this._paddingLeft + this._paddingRight;
  }

  get _paddingLeft() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }

  get _paddingRight() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }
}

export default withExtensions(Button);
