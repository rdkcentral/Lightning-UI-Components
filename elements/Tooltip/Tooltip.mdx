import { Canvas, Story } from '@storybook/addon-docs/blocks';

# Tooltip

Component to provide hint information to other components like buttons and action items.

## Usage

`Tooltip` can display a `title` and is added as a child to any component that needs hint text.  It is positioned above the parent component using the property 'bottomMargin' and centered horizontally based on the parent's width:

```js
import { Tooltip } from '@lightning/ui';
import ActionButton from '../ActionButton';

class Example extends lng.Component {
  static _template() {
    return {
        Button: {
          type: ActionButton,
          x: 200,
          y: 100,
          w: 410,
          title: '+',
          background: 'stroke',
          Tooltip: {
            type: Tooltip,
            title: 'Add to My List!',
            bottomMargin: 24
          }
        }
    };
  }
};
```

<Canvas>
  <Story id="elements-tooltip--basic" />
</Canvas>

The default 'bottomMargin' is 40px above the parent

### Sizing

The size of the tooltip is based on the length of it's 'title'.  If the 'title' exceeds 400px the text wraps, and the background resizes to accomodate.

### Focus

Tooltip is only visible on focus of the parent.

## API

### Properties

name|type|readonly|description
--|--|--|--
title|string|false|tool tip text goes here
bottomMargin|int|false|space between parent component and tooltip in pixels (default: 40px)

