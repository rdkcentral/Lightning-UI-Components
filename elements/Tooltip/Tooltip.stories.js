import lng from '@lightningjs/core';

import Tooltip from '.';
import mdx from './Tooltip.mdx';
import ActionButton from '../ActionButton';

export default {
  title: 'Elements / Tooltip',
  args: {
    focused: true,
    title: 'Tooltip',
    delayVisible: 0,
    timeVisible: 0
  },
  argTypes: {
    focused: { control: 'boolean' },
    title: { control: 'text' },
    delayVisible: { control: 'number' },
    timeVisible: { control: 'number' }
  },
  parameters: {
    argActions: {
      focused: (isFocused, component) => {
        component._getFocused = isFocused
          ? () => component.tag('Tooltip')
          : () => {};
        component._refocus();
      }
    },
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Tooltip: {
          type: Tooltip
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('Tooltip');
      }
    }
  };

export const LongTitle = Basic.bind({});
LongTitle.args = {
  x: 1280 / 2,
  title:
    'This is a long message. Text will remain on a single line and does not have a maximum width'
};

export const WithActionButton = args =>
  class WithActionButton extends lng.Component {
    static _template() {
      return {
        Button: {
          type: ActionButton,
          x: 200,
          y: 100,
          title: 'Button',
          backgroundType: 'stroke',
          Tooltip: {
            type: Tooltip,
            title: args.title,
            delayVisible: args.delayVisible,
            timeVisible: args.timeVisible
          }
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('Button.Tooltip');
      }
    }
  };

WithActionButton.parameters = {
  argActions: {
    focused: (isFocused, component) => {
      component._getFocused = isFocused
        ? () => component.tag('Button.Tooltip')
        : () => {};
      component._refocus();
    },
    title: (title, component) => {
      component.tag('Button.Tooltip').title = title;
    },
    delayVisible: (delayVisible, component) => {
      component.tag('Button.Tooltip').delayVisible = delayVisible;
    },
    timeVisible: (timeVisible, component) => {
      component.tag('Button.Tooltip').timeVisible = timeVisible;
    }
  }
};
