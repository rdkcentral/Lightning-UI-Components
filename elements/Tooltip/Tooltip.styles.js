export default theme => ({
  shadow: theme.materials.glow,
  horizontalPadding: theme.spacing(2),
  radius: theme.border.radius.small,
  textProperties: {
    ...theme.typography.body3,
    textColor: theme.palette.text.dark.primary,
    textAlign: 'center',
    wordWrapWidth: 400
  },
  bottomMargin: theme.spacing(5)
});
