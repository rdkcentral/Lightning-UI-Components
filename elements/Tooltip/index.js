import lng from '@lightningjs/core';
import withStyles from '../../mixins/withStyles';
import withUpdates from '../../mixins/withUpdates';
import { RoundRect } from '../../utils';

export const styles = theme => ({
  shadow: theme.materials.glow,
  horizontalPadding: theme.spacing(2),
  radius: theme.border.radius.small,
  textProperties: {
    ...theme.typography.body3,
    textColor: theme.palette.text.dark.primary,
    textAlign: 'center',
    wordWrapWidth: 400
  },
  bottomMargin: theme.spacing(5)
});

export class Tooltip extends lng.Component {
  static _template() {
    return {
      alpha: 0,
      scale: 0.5,
      mountX: 0.5,
      x: w => w / 2,
      Background: {
        zIndex: 2
      },
      Text: {
        zIndex: 3
      }
    };
  }

  static get properties() {
    return [
      'bottomMargin',
      'horizontalPadding',
      'radius',
      'shadow',
      'textProperties',
      'title'
    ];
  }

  _construct() {
    this.shadow = this.styles.shadow;
    this.textProperties = this.styles.textProperties;
    this.radius = this.styles.radius;
    this.horizontalPadding = this.styles.horizontalPadding;
    this.bottomMargin = this.styles.bottomMargin;
  }

  _init() {
    this._Text.on('txLoaded', () => this._updateContainer());
  }

  _update() {
    this._updateText();
  }

  _updateText() {
    this._Text.text = {
      ...this.textProperties,
      text: this.title
    };
    this._Text.x = this.horizontalPadding;
  }

  _updateContainer() {
    // set background size based on text size
    const backgroundW = this._Text.finalW + this.horizontalPadding * 2;
    const backgroundH =
      this._Text.finalH -
      (this._Text.finalH % this.textProperties.lineHeight) / 2; // calculates even padding

    this.h = backgroundH;
    this.w = backgroundW;

    this._updateBackground();
    this._updateDropShadow();

    this.y = -backgroundH - this.bottomMargin;
  }

  _updateBackground() {
    this._Background.texture = lng.Tools.getRoundRect(
      this.w,
      this.h,
      this.radius
    );
  }

  _updateDropShadow() {
    const DropShadow = this.styles.shadow({
      w: this.w,
      h: this.h,
      borderRadius: this.radius
    });
    this.patch({ DropShadow });
    const alpha = Number(this.hasFocus());
    if (this._smooth) {
      this._DropShadow.smooth = { alpha };
    } else {
      this._DropShadow.alpha = alpha;
    }
  }

  _focus() {
    clearTimeout(this._hideTimer);
    clearTimeout(this._showTimer);

    this.delayVisible
      ? (this._showTimer = setTimeout(() => {
          this.smooth = { alpha: 1, scale: 1 };
        }, this.delayVisible))
      : (this.smooth = { alpha: 1, scale: 1 });

    this._hideTimer = this.timeVisible
      ? setTimeout(() => {
          this._unfocus();
        }, this.timeVisible + this.delayVisible)
      : null;
  }

  _unfocus() {
    clearTimeout(this._hideTimer);
    clearTimeout(this._showTimer);

    this.smooth = { alpha: 0, scale: 0.5 };
  }

  get _Text() {
    return this.tag('Text');
  }
  get _DropShadow() {
    return this.tag('DropShadow');
  }
  get _Background() {
    return this.tag('Background');
  }
}

export default withUpdates(withStyles(Tooltip, styles));
