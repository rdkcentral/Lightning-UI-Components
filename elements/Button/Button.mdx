import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import Button from '.';

# Button

Basic component that handles styling and functionality for a button.

## Usage

Used to convey an interactable portion of the interface, a button can present text, an icon, a checkbox, or both text and an icon or checkbox. It has support for a focused, unfocused, and disabled states.

```js
import { Button } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Button: {
        type: Button,
        title: 'Button'
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-button-themed--basic" />
</Canvas>

### Sizing

By default, `Button` widths dynamically adjust to account for `title` and `icon` widths and `padding`, with a minimum width defined by `minWidth`.

If you want a specific `Button` width no matter what the size of content, set the `fixed` property to `true`. The `minWidth` here will still be honored.

If you want a `Button` that is fully dynamic, set `minWidth` to `1`.

If you'd like to have a fixed button with a width smaller than the `minWidth`, you'll need to set the `minWidth` to a value less than or equal to the provided width.

### Focus

When focused, the background/foreground colors are inverted by default. This can be overridden with theming.

### Prefix

All buttons support two types of Prefixes, a checkbox or an icon.

### Icon

If the button is provided an `icon`, it will be positioned to the left of the title and centered it horizontally and vertically.

### Checkbox

If the button `checkbox` property is set to `true`, a `checkbox` will be positioned to the left of the title and centered it horizontally and vertically.

## API

### Properties

| name     | type    | readonly | default  | description                                                                                |
| -------- | ------- | -------- | -------- | ------------------------------------------------------------------------------------------ |
| checkbox | object  | false    | -        | if defined, a checkbox will be rendered. contains the properties `checked` and `disabled`. |
| disabled | boolean | false    | false    | controls the disabled/enabled state of the button                                          |
| fixed    | boolean | false    | false    | when true, `w` overrides dynamically calclulated width                                     |
| icon     | string  | false    | -        | URI for the icon asset                                                                     |
| justify  | string  | false    | 'center' | alignment of the button's content                                                          |
| title    | string  | false    | -        | button text                                                                                |
| w        | number  | false    | -        | width of the button, if `fixed` is true                                                    |

### Style Properties

| name               | type                          | description                                                                            |
| ------------------ | ----------------------------- | -------------------------------------------------------------------------------------- |
| fontOffsetY        | number                        | used to vertically align the content in the container due to actual font sizing issues |
| justify            | 'center' \| 'left' \| 'right' | alignment of the button's content                                                      |
| minWidth           | number                        | used as a fallback width if no width is passed into the component and `fixed` is false |
| paddingX           | number                        | space between the button horizontal edge and the content                               |
| paddingXIcon       | number                        | space between the button horizontal edge and the content if no title is provided       |
| paddingY           | number                        | space between the button vertical edge and the content                                 |
| prefixMargin       | number                        | the space between the prefix and the button title                                      |
| prefixStyle        | object                        | styles that will be applied to the prefix(icon or checkbox)                            |
| radius             | number                        | shape of component                                                                     |
| textColorFocused   | string                        | color of the title text when the component is focused                                  |
| textColorUnfocused | string                        | color of the title text when the component is not focused                              |
| textColorDisabled  | string                        | color of the title text when the component is disabled                                 |
| textStyle          | string                        | font properties to apply to the title                                                  |
