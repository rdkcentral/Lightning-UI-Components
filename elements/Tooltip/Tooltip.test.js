'use strict';

import TestUtils from '../../test/lightning-test-utils';
import Tooltip from '.';
import Button from '../Button';

const createTooltip = TestUtils.makeCreateComponent(
  Tooltip,
  {},
  { focused: false }
);
const createButton = TestUtils.makeCreateComponent(Button);

describe('Tooltip', () => {
  let tooltip, testRenderer;

  beforeEach(() => {
    [tooltip, testRenderer] = createTooltip({ title: 'Tooltip' });
  });
  afterEach(() => {
    tooltip = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  describe('visibility', () => {
    afterEach(() => {
      jest.clearAllTimers();
      jest.useRealTimers();
    });

    it('should be visible when focused', () => {
      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);
    });

    it('should be hidden when unfocused', () => {
      tooltip._unfocus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(0);
    });

    it('should show after delay if delayVisible is set', () => {
      jest.useFakeTimers();
      expect(tooltip.alpha).toEqual(0);

      // set delayVisible
      tooltip.delayVisible = 5000;

      // should remain hidden
      tooltip._focus();
      testRenderer.update();
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(tooltip.alpha).toEqual(0);

      // should show after timer runs
      jest.runAllTimers();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);
    });

    it('should hide after timer ends if timeVisible is set', () => {
      jest.useFakeTimers();

      // timer not run unless focused
      expect(setTimeout).toHaveBeenCalledTimes(0);

      // show tooltip on focus
      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // check that no timer has been called and tooltip remains visible
      jest.runAllTimers();
      expect(setTimeout).toHaveBeenCalledTimes(0);
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // set timeVisible
      tooltip.timeVisible = 5000;

      // timer not run unless focused
      expect(setTimeout).toHaveBeenCalledTimes(0);

      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // timer started on focus and hide after timer is called
      jest.runAllTimers();
      expect(setTimeout).toHaveBeenCalledTimes(1);
      testRenderer.update();
      expect(tooltip.alpha).toEqual(0);
    });
  });

  describe('position', () => {
    let button, buttonRenderer;
    beforeEach(() => {
      tooltip._focus();

      // eslint-disable-next-line no-unused-vars
      [button, buttonRenderer] = createButton({ title: 'Button' });
      buttonRenderer.update();
    });
    it('should be horizontally centered', () => {});
    it('should default to 40px above parent', done => {
      tooltip._Text.on('txLoaded', () => {
        const buffer = 2;
        expect(tooltip.y).toEqual(-tooltip._Background.finalH - 40 + buffer);

        done();
      });
    });

    it('should change when "bottomMargin" is set', done => {
      const [closer_tooltip] = createTooltip({
        title: 'Tooltip',
        bottomMargin: 20
      });
      closer_tooltip._focus();

      tooltip._Text.on('txLoaded', () => {
        closer_tooltip._Text.on('txLoaded', () => {
          expect(closer_tooltip.y).toEqual(tooltip.y + 20);
          done();
        });
      });
    });
  });

  describe('dimensions', () => {
    // NOTE: the font doesn't text size and calculated tooltip size is not accurate

    beforeEach(() => {
      tooltip._focus();
    });

    it('should be slightly larger than text dimensions', done => {
      testRenderer.update();

      setTimeout(() => {
        try {
          const textWidth = tooltip._Text.finalW;
          const textHeight = tooltip._Text.finalH;

          const diffW = Math.abs(tooltip._Text.finalW - textWidth);
          expect(diffW).toBeLessThan(5);
          expect(tooltip._Background.finalW).toBeGreaterThan(textWidth);

          const diffH = Math.abs(tooltip._Background.finalH - textHeight);
          expect(diffH).toBeLessThan(5);

          done();
        } catch (error) {
          done(error);
        }
      }, 1000);
    });

    it('should be narrower when the message text changes to a shorter message', done => {
      testRenderer.update();

      const controlW = tooltip.finalW;
      const controlBackgroundW = tooltip._Background.finalW;

      tooltip.title = 'short';

      tooltip._Text.loadTexture();
      tooltip._Background.loadTexture();
      testRenderer.update();

      setTimeout(() => {
        try {
          expect(tooltip.finalW).toBeLessThan(controlW);
          expect(tooltip._Background.finalW).toBeLessThan(controlBackgroundW);
          done();
        } catch (error) {
          done(error);
        }
      }, 1000);
    });

    it('should have a longer hint message when the text changes', done => {
      testRenderer.update();

      const controlW = tooltip.finalW;
      const controlBackgroundW = tooltip._Background.finalW;

      tooltip.title = 'a longer title';

      tooltip._Text.loadTexture();
      tooltip._Background.loadTexture();
      testRenderer.update();

      setTimeout(() => {
        try {
          expect(tooltip.finalW).toBeGreaterThan(controlW);
          expect(tooltip._Background.finalW).toBeGreaterThan(
            controlBackgroundW
          );
          done();
        } catch (error) {
          done(error);
        }
      }, 1000);
    });

    it('should change height and wrap text when hint message is long', done => {
      testRenderer.update();

      const controlH = tooltip._Text.finalH;
      const controlBackgroundH = tooltip._Background.finalH;

      tooltip.title =
        'This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...';

      tooltip._Text.loadTexture();
      tooltip._Background.loadTexture();
      testRenderer.update();

      setTimeout(() => {
        try {
          const textHeight = tooltip._Text.finalH;
          const backgroundHeight = tooltip._Background.finalH;

          const diffH = Math.abs(tooltip._Background.finalH - textHeight);

          expect(diffH).toBeLessThan(5);

          expect(textHeight).toBeGreaterThan(controlH);
          expect(backgroundHeight).toBeGreaterThan(controlBackgroundH);

          done();
        } catch (error) {
          done(error);
        }
      }, 1000);
    });
  });
});
