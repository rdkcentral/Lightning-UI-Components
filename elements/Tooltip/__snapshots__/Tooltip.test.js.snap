// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
Object {
  "Component": Object {
    "Background": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Text": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Text": Object {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Text",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Text",
              ],
              "texture": Object {
                "fontFace": "DMSansBold",
                "fontSize": 20,
                "lineHeight": 24,
                "maxLines": undefined,
                "maxLinesSuffix": "...",
                "text": "Tooltip",
                "textBaseline": "bottom",
                "textColor": 4279505943,
                "textOverflow": "ellipsis",
                "type": "TextTexture",
                "verticalAlign": "middle",
                "wordWrapWidth": undefined,
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 2,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 24,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Text",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 7,
          "x": 19.5,
          "y": 20,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 52,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Background",
      ],
      "texture": Object {
        "h": 52,
        "type": "Bubble",
        "w": 39,
      },
      "visible": true,
      "w": 39,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
